broadinstitute/single_cell_portal_core
# Mitigate lag on hovering and scrolling on legends (SCP-4604)

Previous PR - https://github.com/broadinstitute/single_cell_portal_core/pull/1638
This ^ _almost_ fixed the lag however, it failed to address the issue of a build up of calls to updating the active trace label. I closed that PR and am opening this one to keep things clean. The solution presented in this PR builds on the timeout solution by using lodashes debounce. A debounced function will execute after a timeout but has the added benefit of being cancelable. This way the calls to update the active label will not only not fire immediately the previous calls will be canceled so there won't be a buildup of updates waiting to execute.

An important note that wasn't made clear in the previous PR also is that this is not a symptom of scrolling it is a symptom of fast mouse events, and scrolling of a very long list of labels just really shows this issue.

To test:

-It's easiest to see on a really big study. I copied the one from Staging here: https://singlecell-staging.broadinstitute.org/single_cell/study/SCP149/5k-unqiue-u19-multi-cell-type-merged-labels#study-visualize. However this study doesn't have overlapping points on the plot so it's also worthwhile testing with a smaller study that will show the hover functionality of bringing hovered traces to the front.
- To see the laggy issue try scrolling up and down in the legend list in the study in staging.
- Pull this branch and start up your local server and navigate to your really big study that has visualizations
- Try scrolling the legend and observe the lagginess is fixed.
- Also try a study that has overlapping points in the plot and observe that hovering still brings the different points to the front - albeit with a small delay now.

Before:

https://user-images.githubusercontent.com/54322292/195887488-b6f92009-3d6d-41b6-ac05-8870590e6bab.mov

AFter:

No lag on long list:
https://user-images.githubusercontent.com/54322292/195887770-91f3600a-e009-4370-9405-140e1173a2b1.mov

Demonstrate the slight delay now on the hover functionality
https://user-images.githubusercontent.com/54322292/195887779-c52749dd-d3a9-4752-934c-b606637e05b9.mov
